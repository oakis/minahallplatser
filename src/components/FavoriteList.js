import _ from 'lodash';
import React, { Component } from 'react';
import { Keyboard, Alert, AsyncStorage } from 'react-native';
import firebase from 'firebase';
import { connect } from 'react-redux';
import { Actions } from 'react-native-router-flux';
import { Container, Content, List, ListItem, Text, Spinner, Icon } from 'native-base';
import { favoriteGet, favoriteDelete } from '../actions';

class FavoriteList extends Component {

	constructor(props) {
		super(props);
		this.favoriteDelete = this.props.favoriteDelete.bind(this);
	}

	componentWillMount() {
		Keyboard.dismiss();
		firebase.auth().onAuthStateChanged((fbUser) => {
			AsyncStorage.getItem('minahallplatser-user').then((dataJson) => {
				const user = JSON.parse(dataJson);
				if (user.uid === fbUser.uid) {
					this.props.favoriteGet(user);
				}
			});
			this.createDataSource(this.props);
		});
	}

	componentWillReceiveProps(nextProps) {
		this.createDataSource(nextProps);
	}

	createDataSource({ favorites }) {
		this.props.favorites = favorites;
	}

	renderRow(favorite) {
		return (
			<ListItem
				iconRight
				button
				onPress={() => Actions.departures(favorite)}
			>
				<Text>
					{favorite.busStop}
				</Text>
				<Icon
					name='ios-remove-circle-outline'
					style={{ color: '#f00' }}
					onPress={() => {
						Alert.alert(
							favorite.busStop,
							`Är du säker att du vill ta bort ${favorite.busStop}?`,
							[
								{ text: 'Avbryt' },
								{ text: 'Ja', onPress: () => favoriteDelete(favorite.id) }
							]
						);
					}}
				/>
			</ListItem>
		);
	}

	renderList() {
		if (this.props.loading) {
			return (
				<Content contentContainerStyle={{ justifyContent: 'center', alignItems: 'center' }}>
					<Spinner color="blue" />
				</Content>
			);
		} else if (this.props.error) {
			return (
				<Content>
					<Text style={{ textAlign: 'center' }}>{this.props.error}</Text>
				</Content>
			);
		} else if (this.props.favorites.length === 0) {
			return (
				<Content>
					<Text style={{ textAlign: 'center' }}>Du har inte sparat några favoriter än.</Text>;
				</Content>
			);
		}

		return ( 
			<Content>
				<List
					dataArray={this.props.favorites}
					renderRow={this.renderRow}
					favoriteDelete={this.props.favoriteDelete}
				/>
			</Content>
		);
	}

	render() {
		return (
			<Container>
				{this.renderList()}
			</Container>
		);
	}
}

const mapStateToProps = state => {
	const favorites = _.values(state.fav.list);
	const { loading, error } = state.fav;
	return { favorites, loading, error };
};

export default connect(mapStateToProps, { favoriteGet, favoriteDelete })(FavoriteList);
