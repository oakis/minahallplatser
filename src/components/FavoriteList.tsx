import _ from 'lodash';
import React, {useEffect, useState} from 'react';
import {
  Keyboard,
  Alert,
  FlatList,
  View,
  AppState,
  TouchableWithoutFeedback,
  ScrollView,
  StyleProp,
  TextProps,
} from 'react-native';
import analytics from '@react-native-firebase/analytics';
import Icon from 'react-native-vector-icons/MaterialIcons';
import {connect} from 'react-redux';
import {
  favoriteDelete,
  clearErrors,
  searchStops,
  searchChanged,
  favoriteCreate,
  getNearbyStops,
  setSetting,
} from '@actions';
import {
  ListItem,
  Message,
  Input,
  ListItemSeparator,
  ListHeading,
  Text,
  Button,
  MiniMenu,
  Popup,
} from '@common';
import {colors, component, metrics} from '@style';
import {CLR_SEARCH, CLR_ERROR, SEARCH_BY_GPS_FAIL} from '@types';
import {store} from '@src/App';
import {track, isAndroid} from '@helpers';
import {Feedback} from '@modals';
import {useNavigation} from '@react-navigation/native';

interface IProps {
  allowedGPS: boolean;
  gpsLoading: boolean;
  searchLoading: boolean;
  clearErrors: () => void;
  searchChanged: (busStop: string) => void;
  searchStops: (busStop: Record<string, string>) => void;
  getNearbyStops: () => void;
  setSetting: (setting: string, value: string | undefined) => void;
  favoriteDelete: (id: string) => void;
  favoriteCreate: (item: Record<string, string>) => void;
  favorites: IStop[];
  stopsNearby: IStop[];
  departureList: IStop[];
  busStop: string;
  error: string;
}

const FavoriteList = (props: IProps): JSX.Element => {
  const navigation = useNavigation();

  const [editing, setEditing] = useState(false);
  const [showHelp, setShowHelp] = useState(false);
  const [miniMenuOpen, setMiniMenuOpen] = useState(false);
  const [feedbackVisible, setFeedbackVisible] = useState(false);
  const [sortingVisible, setSortingVisible] = useState(false);

  useEffect(() => {
    navigation.setOptions({
      title: 'Mina Hållplatser',
      headerRight: () => (
        <TouchableWithoutFeedback onPress={toggleMiniMenu}>
          <View
            style={{
              width: 30,
              height: 30,
              alignItems: 'center',
              justifyContent: 'center',
              right: 5,
            }}>
            <Icon
              name="more-horiz"
              style={{color: colors.alternative, fontSize: 24}}
            />
          </View>
        </TouchableWithoutFeedback>
      ),
      headerTitleStyle: {
        width: '100%',
        marginHorizontal: 'auto',
        left: 28,
        alignSelf: 'center',
        textAlign: 'center',
        fontSize: 14,
        fontFamily: isAndroid() ? 'sans-serif' : 'System',
      },
    });
    analytics().logScreenView({
      screen_name: 'Dashboard',
      screen_class: 'Dashboard',
    });
    Keyboard.dismiss();
    if (props.allowedGPS) {
      window.log('Refreshing nearby stops');
      props.getNearbyStops();
    }
    track('Page View', {Page: 'Dashboard'});
    AppState.addEventListener('change', handleAppStateChange);
    () => {
      // fetch.abort('searchStops');
      props.clearErrors();
      AppState.removeEventListener('change', handleAppStateChange);
    };
    //eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const onInputChange = (busStop: string) => {
    // fetch.abort('searchStops');
    // clearTimeout(searchTimeout);
    props.searchChanged(busStop);
    // searchTimeout = setTimeout(() => {
    props.searchStops({busStop});
    // }, 500);
  };

  const handleAppStateChange = (nextAppState: string) => {
    if (nextAppState === 'active') {
      if (props.allowedGPS) {
        props.getNearbyStops();
      }
      track('Page View', {Page: 'Dashboard', Parent: 'Background'});
    }
  };

  const resetSearch = () => {
    window.log('resetSearch()');
    store.dispatch({type: CLR_SEARCH});
    store.dispatch({type: CLR_ERROR});
  };

  const refreshNearbyStops = () => {
    track('Refresh NearbyStops');
    store.dispatch({type: SEARCH_BY_GPS_FAIL});
    props.getNearbyStops();
  };

  const toggleMiniMenu = () => {
    setMiniMenuOpen(prev => !prev);
  };

  const closeMiniMenu = () => {
    setMiniMenuOpen(false);
  };

  const renderMiniMenu = () => {
    return (
      <MiniMenu
        isVisible={miniMenuOpen}
        onClose={closeMiniMenu}
        items={[
          {
            icon: 'sort',
            content: 'Sortera favoriter',
            onPress: openSorting,
          },
          {
            icon: 'help',
            content: 'Hjälp',
            onPress: openPopup,
          },
          {
            icon: 'feedback',
            content: 'Lämna feedback',
            onPress: openFeedback,
          },
        ]}
      />
    );
  };

  const openSorting = () => {
    closeMiniMenu();
    setTimeout(() => {
      setSortingVisible(true);
    }, 1);
  };

  const onOrderValueChange = (itemValue: string | undefined) => {
    props.setSetting('favoriteOrder', itemValue);
    setSortingVisible(false);
  };

  const openFeedback = () => {
    track('Feedback Open');
    setFeedbackVisible(true);
    closeMiniMenu();
  };

  const closeFeedback = () => {
    setFeedbackVisible(false);
  };

  const openPopup = () => {
    track('Show Help', {Page: 'Dashboard'});
    closeMiniMenu();
    setTimeout(() => {
      setShowHelp(true);
    }, 1);
  };

  const closeSorting = () => {
    setSortingVisible(false);
  };

  const renderSorting = () => {
    return (
      <MiniMenu
        isVisible={sortingVisible}
        onClose={closeSorting}
        items={[
          {
            content: 'Ingen sortering',
            onPress: () => onOrderValueChange(undefined),
          },
          {
            content: 'Mina mest använda',
            onPress: () => onOrderValueChange('opened'),
          },
          {
            content: 'Efter bokstav',
            onPress: () => onOrderValueChange('busStop'),
          },
        ]}
      />
    );
  };

  const closeHelp = () => {
    setShowHelp(false);
  };

  const renderPopup = () => {
    return (
      <Popup onPress={closeHelp} isVisible={showHelp}>
        <Text style={component.popup.header as StyleProp<TextProps>}>
          Söka efter hållplats
        </Text>
        <Text style={component.popup.text}>
          För att söka på en hållplats klickar du på sökfältet ({' '}
          <Icon name="search" /> ) högst upp på startsidan och fyller i ett
          eller flera sökord.
        </Text>

        <Text style={component.popup.header as StyleProp<TextProps>}>
          Hållplatser nära dig
        </Text>
        <Text style={component.popup.text}>
          Hållplatser som är i din närhet kommer automatiskt att visas sålänge
          du har godkänt att appen får använda din{' '}
          <Text style={{fontWeight: 'bold'}}>plats</Text>. Om du har nekat
          tillgång så kan du klicka på pilen ( <Icon name="refresh" /> ) till
          höger om &quot;Hållplatser nära dig&quot; och godkänna åtkomst till
          platstjänster.
        </Text>

        <Text style={component.popup.header as StyleProp<TextProps>}>
          Spara hållplats som favorit
        </Text>
        <Text style={component.popup.text}>
          Längst till höger på hållplatser nära dig eller i sökresultaten finns
          det en stjärna ( <Icon name="star-border" color={colors.warning} /> ),
          klicka på den för att spara hållplatsen som favorit. Nu kommer
          stjärnan ( <Icon name="star" color={colors.warning} /> ) att bli fylld
          med <Text style={{color: colors.warning}}>orange</Text> färg och
          hållplatsen sparas i listan &quot;Mina Hållplatser&quot;.
        </Text>

        <Text style={component.popup.header as StyleProp<TextProps>}>
          Ta bort hållplats från favoriter
        </Text>
        <Text style={component.popup.text}>
          För att ta bort en hållplats från favoriter så klickar du på{' '}
          <Text style={{fontWeight: 'bold'}}>pennan</Text> ({' '}
          <Icon name="edit" /> ) och sedan på{' '}
          <Text style={{fontWeight: 'bold'}}>minustecknet</Text> ({' '}
          <Icon name="remove-circle-outline" color={colors.danger} /> ) bredvid
          den hållplatsen du vill ta bort.
        </Text>

        <Text style={component.popup.header as StyleProp<TextProps>}>
          Sortera favoriter
        </Text>
        <Text style={component.popup.text}>
          I <Text style={{fontWeight: 'bold'}}>menyn</Text> ({' '}
          <Icon name="menu" /> ) kan du hitta olika sorteringsalternativ, t.ex
          dina mest använda hållplatser.
        </Text>
      </Popup>
    );
  };

  const renderFavoriteItem = ({item}: {item: IStop}) => {
    return (
      <ListItem
        text={item.busStop}
        icon="remove-circle-outline"
        pressItem={async () => {
          Keyboard.dismiss();
          await props.clearErrors();
          navigation.navigate('Departures', {
            busStop: item.busStop,
            id: item.id,
            title: item.busStop,
            parent: 'favorites',
          });
        }}
        pressIcon={() => {
          Keyboard.dismiss();
          Alert.alert(
            item.busStop,
            `Är du säker att du vill ta bort ${item.busStop}?`,
            [
              {text: 'Avbryt'},
              {
                text: 'Ja',
                onPress: () => {
                  track('Favorite Stop Remove', {
                    Stop: item.busStop,
                    Parent: 'Favorite List',
                  });
                  props.favoriteDelete(item.id);
                },
              },
            ],
          );
        }}
        iconVisible={editing}
        iconColor={colors.danger}
      />
    );
  };

  const renderSearchItem = ({item}: {item: IStop}, parent: string) => {
    return (
      <ListItem
        text={item.name}
        icon={item.icon}
        pressItem={() => {
          Keyboard.dismiss();
          navigation.navigate('Departures', {
            busStop: item.name,
            id: item.id,
            title: item.name,
            parent,
          });
        }}
        pressIcon={() => {
          Keyboard.dismiss();
          if (item.icon === 'star') {
            track('Favorite Stop Remove', {
              Stop: item.name,
              Parent: item.parent,
            });
            props.favoriteDelete(item.id);
          } else {
            track('Favorite Stop Add', {Stop: item.name, Parent: item.parent});
            props.favoriteCreate({busStop: item.name, id: item.id});
          }
        }}
        iconVisible
        iconColor={colors.warning}
      />
    );
  };

  const renderNearbyStops = () => {
    return (
      <View>
        <ListHeading
          text="Hållplatser nära dig"
          icon="refresh"
          onPress={refreshNearbyStops}
          loading={props.gpsLoading}
        />
        {!props.gpsLoading && props.stopsNearby.length === 0 ? (
          <Text
            style={{
              marginTop: metrics.margin.md,
              marginLeft: metrics.margin.md,
            }}>
            Vi kunde inte hitta några hållplatser nära dig.
          </Text>
        ) : null}
        <FlatList
          data={props.stopsNearby}
          renderItem={item => renderSearchItem(item, 'nearby stops')}
          keyExtractor={item => item.id}
          ItemSeparatorComponent={ListItemSeparator}
          scrollEnabled={false}
          keyboardShouldPersistTaps="always"
        />
      </View>
    );
  };

  const renderSectionList = () => {
    return (
      <View>
        {props.departureList.length > 0 ? (
          <ListHeading text="Sökresultat" />
        ) : null}
        <FlatList
          data={props.departureList}
          renderItem={item => renderSearchItem(item, 'search')}
          keyExtractor={item => item.id}
          ItemSeparatorComponent={ListItemSeparator}
          scrollEnabled={false}
          keyboardShouldPersistTaps="always"
        />
        {renderNearbyStops()}
        <ListHeading
          text="Mina hållplatser"
          icon={props.favorites.length > 0 ? 'edit' : null}
          iconSize={20}
          onPress={() => {
            track('Edit Stops Toggle', {On: !editing});
            setEditing(prev => !prev);
          }}
        />
        <FlatList
          data={props.favorites}
          renderItem={renderFavoriteItem}
          keyExtractor={item => item.id}
          ItemSeparatorComponent={ListItemSeparator}
          scrollEnabled={false}
          keyboardShouldPersistTaps="always"
          extraData={editing}
        />
      </View>
    );
  };

  const onFocus = () => {
    track('Search Focused');
  };

  return (
    <View style={{flex: 1, backgroundColor: colors.background}}>
      <Feedback visible={feedbackVisible} close={closeFeedback} />
      {renderPopup()}
      {renderSorting()}
      {renderMiniMenu()}
      <ScrollView scrollEnabled keyboardShouldPersistTaps="always">
        <Input
          placeholder="Sök hållplats.."
          onChangeText={onInputChange}
          value={props.busStop}
          icon="search"
          loading={props.searchLoading && props.busStop.length > 0}
          iconRight={props.busStop.length > 0 ? 'close' : null}
          iconRightPress={resetSearch}
          underlineColorAndroid="#fff"
          onFocus={onFocus}
          style={[
            {
              borderRadius: 15,
              paddingLeft: metrics.padding.sm,
              paddingRight: metrics.padding.sm,
              margin: metrics.margin.md,
              backgroundColor: '#fff',
            },
            !isAndroid()
              ? {
                  paddingTop: metrics.padding.md,
                  paddingBottom: metrics.padding.md,
                }
              : null,
          ]}
        />
        {props.error ? (
          <Message
            type="warning"
            message={props.error}
            backgroundColor={colors.warning}
          />
        ) : null}
        {renderSectionList()}
        {props.favorites.length === 0 ? (
          <View
            style={{
              marginTop: metrics.margin.md,
              marginLeft: metrics.margin.md,
              marginRight: metrics.margin.md,
            }}>
            <Text style={{marginBottom: metrics.margin.md}}>
              Du har inte sparat några favoriter än.
            </Text>
            <Button
              onPress={openPopup}
              label="Hjälp"
              icon="live-help"
              color={colors.primary}
            />
          </View>
        ) : null}
      </ScrollView>
    </View>
  );
};

const mapStateToProps = (state: IStateProps) => {
  const {favoriteOrder, allowedGPS} = state.settings;
  const favorites = _.orderBy(
    state.fav.favorites,
    o => o[favoriteOrder] || 0,
    favoriteOrder === 'busStop' ? 'asc' : 'desc',
  );
  const {error} = state.errors;
  const favoriteIds = _.map(state.fav.favorites, 'id');
  const {busStop, stops, gpsLoading} = state.search;
  const stopsNearby = _.map(stops, item => {
    return {
      ...item,
      icon: _.includes(favoriteIds, item.id) ? 'star' : 'star-border',
      parent: 'Stops Nearby',
    };
  });
  const searchLoading = state.search.loading;
  const departureList = _.map(state.search.departureList, item => {
    return {
      ...item,
      icon: _.includes(favoriteIds, item.id) ? 'star' : 'star-border',
      parent: 'Search List',
    };
  });
  return {
    favorites,
    error,
    busStop,
    departureList,
    favoriteIds,
    searchLoading,
    stopsNearby,
    gpsLoading,
    allowedGPS,
  };
};

export default connect(mapStateToProps, {
  favoriteDelete,
  clearErrors,
  searchStops,
  searchChanged,
  favoriteCreate,
  getNearbyStops,
  setSetting,
})(FavoriteList);
