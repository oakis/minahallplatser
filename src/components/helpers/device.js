import { Platform } from 'react-native';

export const isAndroid = () => {
    return Platform.OS === 'android';
};

export const isIOS = () => {
    return Platform.OS === 'ios';
};
