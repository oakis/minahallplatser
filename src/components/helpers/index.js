export * from './device';
export * from './message';
export * from './network';
export * from './token';
export * from './firebase';
export * from './analytics';
export * from './globals';
export * from './image';

export const isObject = (obj) => {
    return Object.prototype.toString.call(obj) === '[object Object]';
};
