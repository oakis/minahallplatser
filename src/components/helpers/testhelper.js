export const navigation = {
    getParam: jest.fn(),
    setParams: jest.fn(),
    navigate: jest.fn(),
};
