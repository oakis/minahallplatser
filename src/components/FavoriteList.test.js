import React from 'react';
import { AppState, Keyboard } from 'react-native';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { stub } from 'sinon';
import fetch from 'react-native-cancelable-fetch';
import FavoriteList from './FavoriteList';
import { track } from './helpers';
import { CLR_ERROR, CLR_SEARCH } from '../actions/types';
import { store } from '../App';
import { navigation } from './helpers/testhelper.js';

const mockStore = configureMockStore([thunk]);
const initialState = {
    settings: {
        favoriteOrder: 'nothing',
        allowedGPS: false,
    },
    fav: {
        favorites: [
            { id: '9021014007900000', busStop: 'Överåsvallen, Göteborg' },
            { id: '9021014016323000', busStop: 'Älmhult, Ale' },
        ],
        lines: [],
    },
    errors: {
        error: 'Felmeddelande'
    },
    search: {
        busStop: '',
        stops: [
            { id: '9021014007483000', busStop: 'Volvo Torslanda PV, Göteborg' },
            { id: '9021014007900000', busStop: 'Överåsvallen, Göteborg' },
        ],
        gpsLoading: false,
        departureList: [
            { id: '9021014007483000', busStop: 'Volvo Torslanda PV, Göteborg' },
            { id: '9021014007900000', busStop: 'Överåsvallen, Göteborg' },
        ],
    }
};

window.log = () => {};

// REWRITE?
it('getNearbyStops should be called if logged in and user has accepted GPS', async () => {
    const getNearbyStops = stub();
    initialState.settings.allowedGPS = true;
    await shallow(<FavoriteList navigation={navigation} store={mockStore(initialState)} getNearbyStops={getNearbyStops} />).dive();
    expect(getNearbyStops.callCount).toBe(1);
});

it('HelpButton to the right side of the navbar, and state.init to be false', () => {
    const wrapper = shallow(<FavoriteList navigation={navigation} store={mockStore(initialState)} getNearbyStops={jest.fn()} />).dive();
    wrapper.setProps({ getNearbyStops: jest.fn() }); // Must set something to actually run.
    expect(wrapper.state().init).toBe(false);
});

it('should match snapshot', () => {
    const wrapper = shallow(<FavoriteList navigation={navigation} store={mockStore(initialState)} getNearbyStops={jest.fn()} />).dive();
    expect(toJson(wrapper)).toMatchSnapshot();
});

describe('onInputChange', () => {
    let wrapper;
    const searchChanged = stub();
    const searchStops = stub();

    beforeEach(() => {
        wrapper = shallow(<FavoriteList navigation={navigation} store={mockStore(initialState)} searchChanged={searchChanged} searchStops={searchStops} getNearbyStops={jest.fn()} />).dive();
        searchChanged.reset();
    });
    it('should abort ongoing searches', () => {
        fetch.abort = stub();
        wrapper.instance().onInputChange('a');
        expect(fetch.abort.callCount).toBe(1);
    });
    it('should reset search timeout', () => {
        wrapper.instance().clearTimeout = stub();
        wrapper.instance().searchTimeout = stub();
        wrapper.instance().onInputChange('a');
        expect(wrapper.instance().clearTimeout).toBe(undefined);
    });

    it('should call searchChanged', () => {
        wrapper.instance().onInputChange('a');
        expect(searchChanged.callCount).toBe(1);
    });

    it('should call searchStops after some time', () => {
        jest.useFakeTimers();
        wrapper.instance().onInputChange('a');
        const timer = wrapper.instance().searchTimeout;
        expect(timer).toBeDefined();
        jest.runAllTimers();
        expect(searchStops.callCount).toBe(1);
    });
});

describe('componentWillUnmount', () => {
    let wrapper;
    const clearErrors = stub();

    beforeEach(() => {
        wrapper = shallow(<FavoriteList navigation={navigation} store={mockStore(initialState)} clearErrors={clearErrors} getNearbyStops={jest.fn()} />).dive();
        clearErrors.reset();
    });
    it('should abort ongoing http requests', () => {
        fetch.abort = stub();
        wrapper.unmount();
        expect(fetch.abort.callCount).toBe(1);
    });

    it('should call abort with "searchStops"', () => {
        fetch.abort = jest.fn();
        wrapper.unmount();
        expect(fetch.abort).toBeCalledWith('searchStops');
    });

    it('should clear all errors', () => {
        wrapper.unmount();
        expect(clearErrors.callCount).toBe(1);
    });

    it('should remove appstate event listener', () => {
        AppState.removeEventListener = stub();
        wrapper.unmount();
        expect(AppState.removeEventListener.callCount).toBe(1);
    });
});

describe('handleAppStateChange', () => {
    let wrapper;
    const getNearbyStops = stub();

    beforeEach(() => {
        wrapper = shallow(<FavoriteList navigation={navigation} store={mockStore(initialState)} getNearbyStops={getNearbyStops} />).dive();
        getNearbyStops.reset();
    });
    it('should refresh nearby stops if prev accepted by user', () => {
        initialState.settings.allowedGPS = true;
        wrapper.instance().handleAppStateChange('active');
        expect(getNearbyStops.callCount).toBe(1);
    });

    it('should track page view if user is reopening app', () => {
        track.mockReset();
        wrapper.instance().handleAppStateChange('active');
        expect(track).toBeCalledWith('Page View', { Page: 'Dashboard', Parent: 'Background' });
    });
});

it('resetSearch should dispatch CLR_SEARCH & CLR_ERROR', () => {
    store.dispatch = jest.fn();
    const wrapper = shallow(<FavoriteList navigation={navigation} store={mockStore(initialState)} getNearbyStops={jest.fn()} />).dive();
    wrapper.instance().resetSearch();
    expect(store.dispatch).toBeCalledWith({ type: CLR_SEARCH });
    expect(store.dispatch).toBeCalledWith({ type: CLR_ERROR });
});

describe('refreshNearbyStops', () => {
    let wrapper;
    const setSetting = jest.fn();
    const getNearbyStops = stub();

    beforeEach(() => {
        wrapper = shallow(<FavoriteList navigation={navigation} store={mockStore(initialState)} setSetting={setSetting} getNearbyStops={getNearbyStops} />).dive();
    });
    it('should be tracked', () => {
        track.mockReset();
        wrapper.instance().refreshNearbyStops();
        expect(track).toBeCalledWith('Refresh NearbyStops');
    });

    it('should call getNearbyStops', () => {
        getNearbyStops.reset();
        wrapper.instance().refreshNearbyStops();
        expect(getNearbyStops.callCount).toBe(1);
    });
});

describe('openPopup', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallow(<FavoriteList navigation={navigation} store={mockStore(initialState)} getNearbyStops={jest.fn()} />).dive();
    });
    it('should be tracked', () => {
        track.mockReset();
        wrapper.instance().openPopup();
        expect(track).toBeCalledWith('Show Help', { Page: 'Dashboard' });
    });

    it('should set showHelp state to true', () => {
        expect(wrapper.state().showHelp).toBe(false);
        wrapper.instance().openPopup();
        expect(wrapper.state().showHelp).toBe(true);
    });
});

describe('renderFavoriteItem', () => {
    let wrapper;
    let ListItem;
    const clearErrors = stub();
    Keyboard.dismiss = stub();

    beforeEach(() => {
        wrapper = shallow(<FavoriteList navigation={navigation} store={mockStore(initialState)} clearErrors={clearErrors} getNearbyStops={jest.fn()} />).dive();
        ListItem = wrapper.find('FlatList').last().props().renderItem({ item: { busStop: 'Centralstationen', id: '1' } });
        Keyboard.dismiss.reset();
        clearErrors.reset();
    });

    describe('pressItem', () => {
        beforeEach(async () => {
            await ListItem.props.pressItem();
        });
        it('should dismiss keyboard', () => {
            expect(Keyboard.dismiss.callCount).toBe(1);
        });

        it('should clear all errors', () => {
            expect(clearErrors.callCount).toBe(1);
        });

        it('should change view to departures', () => {
            expect(navigation.navigate).toBeCalledWith('Departures',
                {
                    busStop: 'Centralstationen',
                    id: '1',
                    title: 'Centralstationen',
                    parent: 'favorites',
                }
            );
        });
    });

    describe('pressIcon', () => {
        beforeEach(() => {
            ListItem.props.pressIcon();
        });
        it('should dismiss keyboard', () => {
            expect(Keyboard.dismiss.callCount).toBe(1);
        });
    });
});
