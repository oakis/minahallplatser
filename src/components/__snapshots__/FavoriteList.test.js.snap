// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#EAE9EF",
      "flex": 1,
    }
  }
>
  <Feedback
    close={[Function]}
    visible={false}
  />
  <Popup
    isVisible={false}
    onPress={[Function]}
  >
    <DefaultFont
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
          "marginBottom": 5,
        }
      }
    >
      Söka efter hållplats
    </DefaultFont>
    <DefaultFont
      style={
        Object {
          "marginBottom": 10,
        }
      }
    >
      För att söka på en hållplats klickar du på sökfältet ( 
      <MaterialIcon
        name="search"
      />
       ) högst upp på startsidan och fyller i ett eller flera sökord.
    </DefaultFont>
    <DefaultFont
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
          "marginBottom": 5,
        }
      }
    >
      Hållplatser nära dig
    </DefaultFont>
    <DefaultFont
      style={
        Object {
          "marginBottom": 10,
        }
      }
    >
      Hållplatser som är i din närhet kommer automatiskt att visas sålänge du har godkänt att appen får använda din 
      <DefaultFont
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        plats
      </DefaultFont>
      . Om du har nekat tillgång så kan du klicka på pilen ( 
      <MaterialIcon
        name="refresh"
      />
       ) till höger om "Hållplatser nära dig" och godkänna åtkomst till platstjänster.
    </DefaultFont>
    <DefaultFont
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
          "marginBottom": 5,
        }
      }
    >
      Spara hållplats som favorit
    </DefaultFont>
    <DefaultFont
      style={
        Object {
          "marginBottom": 10,
        }
      }
    >
      Längst till höger på hållplatser nära dig eller i sökresultaten finns det en stjärna ( 
      <MaterialIcon
        color="#FFB347"
        name="star-border"
      />
       ), klicka på den för att spara hållplatsen som favorit. Nu kommer stjärnan ( 
      <MaterialIcon
        color="#FFB347"
        name="star"
      />
       ) att bli fylld med 
      <DefaultFont
        style={
          Object {
            "color": "#FFB347",
          }
        }
      >
        orange
      </DefaultFont>
       färg och hållplatsen sparas i listan "Mina Hållplatser".
    </DefaultFont>
    <DefaultFont
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
          "marginBottom": 5,
        }
      }
    >
      Ta bort hållplats från favoriter
    </DefaultFont>
    <DefaultFont
      style={
        Object {
          "marginBottom": 10,
        }
      }
    >
      För att ta bort en hållplats från favoriter så klickar du på 
      <DefaultFont
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        pennan
      </DefaultFont>
       ( 
      <MaterialIcon
        name="edit"
      />
       ) och sedan på 
      <DefaultFont
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        minustecknet
      </DefaultFont>
       ( 
      <MaterialIcon
        color="#FF6961"
        name="remove-circle-outline"
      />
       ) bredvid den hållplatsen du vill ta bort.
    </DefaultFont>
    <DefaultFont
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
          "marginBottom": 5,
        }
      }
    >
      Sortera favoriter
    </DefaultFont>
    <DefaultFont
      style={
        Object {
          "marginBottom": 10,
        }
      }
    >
      I 
      <DefaultFont
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        menyn
      </DefaultFont>
       ( 
      <MaterialIcon
        name="menu"
      />
       ) kan du hitta olika sorteringsalternativ, t.ex dina mest använda hållplatser.
    </DefaultFont>
  </Popup>
  <MiniMenu
    isVisible={false}
    items={
      Array [
        Object {
          "content": "Ingen sortering",
          "onPress": [Function],
        },
        Object {
          "content": "Mina mest använda",
          "onPress": [Function],
        },
        Object {
          "content": "Efter bokstav",
          "onPress": [Function],
        },
      ]
    }
    onClose={[Function]}
  />
  <MiniMenu
    isVisible={false}
    items={
      Array [
        Object {
          "content": "Sortera favoriter",
          "icon": "sort",
          "onPress": [Function],
        },
        Object {
          "content": "Hjälp",
          "icon": "help",
          "onPress": [Function],
        },
        Object {
          "content": "Lämna feedback",
          "icon": "feedback",
          "onPress": [Function],
        },
      ]
    }
    onClose={[Function]}
  />
  <ScrollView
    keyboardShouldPersistTaps="always"
    scrollEnabled={true}
  >
    <Input
      icon="search"
      iconRight={null}
      iconRightPress={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="Sök hållplats.."
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
            "borderRadius": 15,
            "margin": 10,
            "paddingLeft": 5,
            "paddingRight": 5,
          },
          Object {
            "paddingBottom": 10,
            "paddingTop": 10,
          },
        ]
      }
      underlineColorAndroid="#fff"
      value=""
    />
    <Message
      message="Felmeddelande"
      type="warning"
    />
    <View>
      <ListHeading
        text="Sökresultat"
      />
      <FlatList
        ItemSeparatorComponent={[Function]}
        data={
          Array [
            Object {
              "busStop": "Volvo Torslanda PV, Göteborg",
              "icon": "star-border",
              "id": "9021014007483000",
              "parent": "Search List",
            },
            Object {
              "busStop": "Överåsvallen, Göteborg",
              "icon": "star",
              "id": "9021014007900000",
              "parent": "Search List",
            },
          ]
        }
        keyExtractor={[Function]}
        keyboardShouldPersistTaps="always"
        renderItem={[Function]}
        scrollEnabled={false}
      />
      <View>
        <ListHeading
          icon="refresh"
          loading={false}
          onPress={[Function]}
          text="Hållplatser nära dig"
        />
        <FlatList
          ItemSeparatorComponent={[Function]}
          data={
            Array [
              Object {
                "busStop": "Volvo Torslanda PV, Göteborg",
                "icon": "star-border",
                "id": "9021014007483000",
                "parent": "Stops Nearby",
              },
              Object {
                "busStop": "Överåsvallen, Göteborg",
                "icon": "star",
                "id": "9021014007900000",
                "parent": "Stops Nearby",
              },
            ]
          }
          keyExtractor={[Function]}
          keyboardShouldPersistTaps="always"
          renderItem={[Function]}
          scrollEnabled={false}
        />
      </View>
      <ListHeading
        icon="edit"
        iconSize={16}
        onPress={[Function]}
        text="Mina hållplatser"
      />
      <FlatList
        ItemSeparatorComponent={[Function]}
        data={
          Array [
            Object {
              "busStop": "Överåsvallen, Göteborg",
              "id": "9021014007900000",
            },
            Object {
              "busStop": "Älmhult, Ale",
              "id": "9021014016323000",
            },
          ]
        }
        extraData={false}
        keyExtractor={[Function]}
        keyboardShouldPersistTaps="always"
        renderItem={[Function]}
        scrollEnabled={false}
      />
    </View>
  </ScrollView>
</View>
`;
